libraries/AP_Avoidance/AP_Avoidance.h

```diff
@@ -0,0 +1,197 @@
+    AP_Int8     _warn_time_horizon_xy;
+    AP_Int8     _warn_time_horizon_z;
+    AP_Int16    _warn_distance_xy;
+    AP_Int16    _warn_distance_z;
+
+protected:
+
+    enum state_t {
+        STATE_CLEAR = 0,
+        STATE_WARN = 1,
+        STATE_FAIL = 2,
+    };
+    // contains English names corresponding to state_t entries
+    static const char *_state_names[];
+
+    uint32_t _last_state_change_ms;
+    // we will not recover from a state any faster than this
+    static const uint8_t _state_recovery_hysteresis = 10; // seconds
+    state_t _old_state = STATE_CLEAR;
+
+    enum avoidance_recovery_f_t {
+        AVOIDANCE_RECOVERY_F_CONTINUE_FAIL,
+        AVOIDANCE_RECOVERY_F_MOVE_TO_WARN,
+    };
+    enum avoidance_recovery_w_t {
+        AVOIDANCE_RECOVERY_W_RESUME,
+        AVOIDANCE_RECOVERY_W_CONTINUE_ACTION
+    };
+
+    uint32_t _gcs_cleared_messages_first_sent;
+    // specifies how long we should continue sending messages about a threat after it has cleared
+    static const uint8_t _gcs_cleared_messages_duration = 5; // seconds
+
+
+    void handle_threat_gcs_notify(AP_Avoidance::Sample *threat);
+
+    AP_Avoidance::Sample *most_serious_threat();
+
+    // reference to AHRS, so we can ask for our position, heading and
+    // speed
+    const AP_AHRS &_ahrs;
+
+    void internal_error();
+
+    // Deal with the most significant threat
+    virtual void handle_avoidance(AP_Avoidance::Sample *threat);
+
+    // returns an entry from the MAV_COLLISION_ACTION representative
+    // of what the curent avoidance handler is up to.
+    MAV_COLLISION_ACTION mav_avoidance_action();
+
+    // returns an object which is responsible for avoiding the most significant threat
+    virtual class AvoidanceHandler &handler_for_action(MAV_COLLISION_ACTION action) = 0;
+
+private:
+
+    class AP_ADSB &_adsb;
+
+    static AP_Avoidance *_instance;
+
+    uint8_t _sample_count = 0;
+    AP_Avoidance::Sample *_samples;
+
+    // check to see if we are initialised (and possibly do initialisation)
+    bool check_startup();
+
+    // initialize _sample_list
+    void init();
+
+    // free _sample_list
+    void deinit();
+
+    const uint32_t MAX_SAMPLE_AGE_MS = 60000;
+//    const uint32_t MAX_SAMPLE_AGE_MS = 1000; 
+
+    uint32_t src_id_for_adsb_vehicle(AP_ADSB::adsb_vehicle_t vehicle) const;
+
+    void check_for_threats();
+    void update_threat_level(const Location &my_loc,
+                             const Vector3f &my_vel,
+                             AP_Avoidance::Sample &sample);
+    void get_adsb_samples();
+
+    const static uint8_t _gcs_notify_interval = 1; // seconds
+
+    // in SITL, at least, the threat level never remains higher for
+    // more than a fraction of a second.  This could potentially lead
+    // to an aircraft switching rapidly between its avoidance
+    // action and its recovery action.
+    const static uint8_t _minimum_handler_duration = 1; // seconds
+
+    int8_t _current_most_serious_threat;
```

Comment:
It is initialized in init(); I think we avoid initializing stuff unnecessarily still?

Also, there's lots more initialised in the same way...

Given we index _obstacles based on this, it really doesn't matter if it is 0 or -1 - if you use it when init() hasn't been passed through *bad* things are going to happen :-)